## <span style="color:blue">Problem Set 9</span>

<Enter your name here>

### <span style="color:red">Run the cell below first</span>
Before starting the assignment, run the cell below. 

In [4]:
import requests
import json

### Overview
In this problem set, you will make use of the OMDB API, which provides information about movies. The instructions will walk you through creating an interactive service that runs inside the Python notebook. It will do the following:
- keeps track of movies the user says they have seen, in a file and in a python list
- asks the user to enter one or more movie titles they are considering watching
- for each movie title the user enters:
  - present some information about the movie, gathered from OMDB
  - if they user says they plan to watch it; add it to the list of already-watched movies.
  
#### Get an OMDB API key
- Get an OMDB API Key from: http://www.omdbapi.com/apikey.aspx 
- Be sure to respond to the email they send you, in order to activate the key
- Fill in the OMDB_API_KEY field below:
- After filling in the key below, run the cell.

In [5]:
OMDB_API_KEY = 'd00b3286'

Below is an example of the json returned by the API.  You will use the highlighted keys and associated values in your code below.

{__'Title'__: 'Black Panther',<br />
__'Year'__: '2018',<br />
 'Rated': 'PG-13',<br />
 'Released': '16 Feb 2018',<br />
 'Runtime': '134 min',<br />
 'Genre': 'Action, Adventure, Sci-Fi',<br />
 'Director': 'Ryan Coogler',<br />
 'Writer': 'Ryan Coogler, Joe Robert Cole, Stan Lee',<br />
 'Actors': "Chadwick Boseman, Michael B. Jordan, Lupita Nyong'o",<br />
 __'Plot'__: "T'Challa, heir to the hidden but advanced kingdom of Wakanda, must step forward to lead his people into a new future and must confront a challenger from his country's past.",<br />
...<br />
 'BoxOffice': '$700,426,566',<br />
 'Production': 'N/A',<br />
 'Website': 'N/A',<br />

### Problem 1
Define a function *getOMDBData* that calls the OMDB API (described here: http://www.omdbapi.com) to get the data about a movie.

Note: You should use the "By ID or Title" features of the OMDB API.  Note, the API parameters are:
- The title parameter is named _t_
- The API key parameter is name _apikey_

Use the URL _http://www.omdbapi.com/_.  

The function *getOMDBData* accepts the parameter _title_ and does the following
1. Try: to retrieve the movie data using the _requests.get_ method using the parameters _t_ and _apikey_.  The call to _requests.get_ should include the _URL_ and the _params_ parameters.  
2.     If data is retrieved, store the json returned into a _variable_. 
3.     If the _Response_ key of the _variable_ has a string value equal to 'False', return a _boolean_ False.  
4.     Otherwise return the _variable_.
5. If the Try fails, return a _boolean_ False.

In [6]:
def getOMDBData(title):
    try:
        params={'t':title, 'apikey':OMDB_API_KEY}
        res=requests.get("http://www.omdbapi.com/", params)
        #print("request success")
        omdb_json=json.loads(res.text)
        if res==False:
            return False
        else:
            return(omdb_json)
    except:
        return False
    
        

In [7]:
# run this cell to test your function with a valid movie title
getOMDBData('Black Panther')

{'Title': 'Black Panther',
 'Year': '2018',
 'Rated': 'PG-13',
 'Released': '16 Feb 2018',
 'Runtime': '134 min',
 'Genre': 'Action, Adventure, Sci-Fi',
 'Director': 'Ryan Coogler',
 'Writer': 'Ryan Coogler, Joe Robert Cole, Stan Lee',
 'Actors': "Chadwick Boseman, Michael B. Jordan, Lupita Nyong'o",
 'Plot': "T'Challa, heir to the hidden but advanced kingdom of Wakanda, must step forward to lead his people into a new future and must confront a challenger from his country's past.",
 'Language': 'English, Swahili, Nama, Xhosa, Korean',
 'Country': 'United States',
 'Awards': 'Won 3 Oscars. 117 wins & 282 nominations total',
 'Poster': 'https://m.media-amazon.com/images/M/MV5BMTg1MTY2MjYzNV5BMl5BanBnXkFtZTgwMTc4NTMwNDI@._V1_SX300.jpg',
 'Ratings': [{'Source': 'Internet Movie Database', 'Value': '7.3/10'},
  {'Source': 'Rotten Tomatoes', 'Value': '96%'},
  {'Source': 'Metacritic', 'Value': '88/100'}],
 'Metascore': '88',
 'imdbRating': '7.3',
 'imdbVotes': '795,068',
 'imdbID': 'tt1825683

<span style="color:blue">Expected output:</span><br/>
<span style="font: courier; font-size: 13px">
{'Actors': "Chadwick Boseman, Michael B. Jordan, Lupita Nyong'o, Danai Gurira",<br />
 'Awards': '14 nominations.',<br />
 'BoxOffice': '$501,105,037',<br />
 'Country': 'USA, South Africa, South Korea, Australia',<br />
 'DVD': '15 May 2018',<br />
 'Director': 'Ryan Coogler',<br />
 'Genre': 'Action, Adventure, Sci-Fi',<br />
 'Language': 'Swahili, Nama, English, Xhosa, Korean',<br />
 _remaining output omitted_</span>

In [8]:
# run this cell to test your function with a invalid movie title
getOMDBData("asdfasdfasdfasdfasdf")

{'Response': 'False', 'Error': 'Movie not found!'}

<span style="color:blue">Expected output:</span><br/>
<span style="font: courier; font-size: 13px">
False</span>

### Problem 2: 
Define the function *getMoviesData* that: 
- accepts a *list* of titles (strings) as input parameter
- returns a *list* of dictionaries fetched from the OMDB API, by calling your getOMDBData function.


In [9]:
def getMoviesData(movieTitles):
    data=[]
    for movie in movieTitles:
        data+=[getOMDBData(movie)]
    return data
    return None

In [10]:
# Run this cell to test the function
movies = getMoviesData(["Black Panther", "Clue","foobarbat"])
print(movies)

[{'Title': 'Black Panther', 'Year': '2018', 'Rated': 'PG-13', 'Released': '16 Feb 2018', 'Runtime': '134 min', 'Genre': 'Action, Adventure, Sci-Fi', 'Director': 'Ryan Coogler', 'Writer': 'Ryan Coogler, Joe Robert Cole, Stan Lee', 'Actors': "Chadwick Boseman, Michael B. Jordan, Lupita Nyong'o", 'Plot': "T'Challa, heir to the hidden but advanced kingdom of Wakanda, must step forward to lead his people into a new future and must confront a challenger from his country's past.", 'Language': 'English, Swahili, Nama, Xhosa, Korean', 'Country': 'United States', 'Awards': 'Won 3 Oscars. 117 wins & 282 nominations total', 'Poster': 'https://m.media-amazon.com/images/M/MV5BMTg1MTY2MjYzNV5BMl5BanBnXkFtZTgwMTc4NTMwNDI@._V1_SX300.jpg', 'Ratings': [{'Source': 'Internet Movie Database', 'Value': '7.3/10'}, {'Source': 'Rotten Tomatoes', 'Value': '96%'}, {'Source': 'Metacritic', 'Value': '88/100'}], 'Metascore': '88', 'imdbRating': '7.3', 'imdbVotes': '795,068', 'imdbID': 'tt1825683', 'Type': 'movie', '

<span style="color:blue">Expected output:</span><br/>
<span style="font: courier; font-size: 13px">
{'Title': 'Black Panther', 'Year': '2018', 'Rated': 'PG-13', 'Released': '16 Feb 2018', 'Runtime': '134 min', 'Genre':<br /> 'Action, Adventure, Sci-Fi', 'Director': 'Ryan Coogler', 'Writer': 'Ryan Coogler, Joe Robert Cole, Stan Lee (based on ... _remaining output omitted_</span>

### Problem 3: 
Define the function *getValidMoviesData* that:
- accepts a *list* of movie titles
- Uses the getMoviesData function to return a list of valid movie dictionaries.

This function needs to test the return value from getMoviesData to filter out any False return values. 

In [11]:
def getValidMoviesData(movieTitles):
    valid_list=[]
    validmovie=''
    
    for movie in movieTitles:
        validmovie=getMoviesData(movieTitles)
    for title in validmovie:
        if title['Response'] !='False':
            valid_list.append(title)
    return valid_list
    return None

In [12]:
# Run this cell to test the function
for movie in getValidMoviesData(["Superbabies: Baby Geniuses 2", 
                  "The Shawshank Redemption", 
                  "Foobarbat",
                  "asdfasdfasdfasdfasdf" , 
                  "Thor"]):
    print(movie['Title'],movie['Year'])

Superbabies: Baby Geniuses 2 2004
The Shawshank Redemption 1994
Thor 2011


<span style="color:blue">Expected output:</span><br/>
<span style="font: courier; font-size: 13px">
Superbabies: Baby Geniuses 2 2004<br/>
The Shawshank Redemption 1994<br/>
Thor 2011</span>

### Problem 4: 
Define a function *sortedMoviesData* that:
- accepts a *list* of movie titles
- builds the list of dictionaries returned by getValidMoviesData 
- returns the list of dictionaries sorted _high to low_ on the IMDB rating

Note: Be sure to cast the ratings as floats.  The optomized solution uses the sorted function and a lambda expression.

In [223]:
def sortedMoviesData(movieTitles):
    validlist=[]
    validlist=getValidMoviesData(movieTitles)
    sortedList=[]
    imdbDict={}
    
    for movie in validlist:
        imdbRating=float(movie['imdbRating'])
        imdbDict[imdbRating]=movie
        pairs=imdbDict.items()
        sortPairs= sorted(pairs, key=lambda pairs:pairs[0], reverse=True ) 
        
    for i in sortPairs:
        sortedList.append(i[1])
    return sortedList
    

In [224]:
result = sortedMoviesData(["Superbabies: Baby Geniuses 2", 
                  "The Shawshank Redemption", 
                  "Clue",
                  "asdfasdfasdfasdfasdf" , 
                  "Black Panther",
                  "Lord of the Rings",
                  "Source Code"])
for movie in result:
    print(movie['Title'],movie['Year'],movie['imdbRating'])

The Shawshank Redemption 1994 9.3
The Lord of the Rings: The Fellowship of the Ring 2001 8.8
Source Code 2011 7.5
Black Panther 2018 7.3
Clue 1985 7.2
Superbabies: Baby Geniuses 2 2004 1.5


<span style="color:blue">Expected output:</span><br/>
<span style="font: courier; font-size: 13px">
The Shawshank Redemption 1994 9.3<br/>
Lord of the Rings 2014 8.8<br/>
Source Code 2011 7.5<br/>
Clue 1985 7.3<br/>
Black Panther 2018 7.3<br/>
Superbabies: Baby Geniuses 2 2004 1.9</span>

### PROBLEM 5:
The next few problems require the use of a file that tracks the movies that the user has viewed.   

Write function called _loadViewedMovies_ that does the following:
1. Open the file *PREVIOUSLY_VIEWED_FNAME* (see below)
2. If the file opens successfully:
   1. Read and load its contents using json into the variable _viewed_movies_.
3. Otherwise:  #this is the case for the first time the cell is run
   1. Assign _viewed_movies_ an empty list.
4. Returns the _viewed_movies_.

Notes:
- Use try/except when opening the file.
- The contents of PREVIOUSLY_VIEWED_FNAME will be a list of movie titles
- If code you will develop in other cells below gets into a bad state, you may have to delete the _viewed_movies.txt_ file and recreate.

In [188]:
PREVIOUSLY_VIEWED_FNAME = 'viewed_movies.txt'

# write your code below
def loadViewedMovies():
    try:
        moviefile=open(PREVIOUSLY_VIEWED_FNAME,'r')
        contents=moviefile.read()
        viewed_movies=json.loads(contents)
    except:
        viewed_movies=[]

        
    
    return viewed_movies

In [189]:
# Run this to test the above code.  The first time this cell is run, an empty list will print.
#  Note:  use viewed_movies in code you will write later
viewed_movies = loadViewedMovies()
print(viewed_movies)

["It's a Beautiful Life", 'Black Panther', 'Thor']


<span style="color:blue">Expected output:</span><br/>
<span style="font: courier; font-size: 13px">
[]<br />
</span>

### PROBLEM 6: 
 Define a function *hasSeenMovie* that:
 - accepts a title 
 - returns True if that title is in the viewed_movies list
 - otherwise returns False


In [125]:
def hasSeenMovie(title):
    if title in viewed_movies:
        return True
    else:
        return False
    return None

### PROBLEM 7:
Define a function *markAsViewed* that accepts a title, and:
1. If that title IS in viewed_movies, do nothing
2. If that title IS NOT in viewed_movies
    - add the title to the list viewed_movies
    - write the entire contents of viewed_movies to the file PREVIOUSLY_VIEWED_FNAME as a json-formatted string
3. return nothing


In [179]:
def markAsViewed(title):
    if hasSeenMovie(title)==True:
        return None
    else:
        viewed_movies.append(title)
        delimiter=" "
        with open(PREVIOUSLY_VIEWED_FNAME, 'w') as outfile:
            string=json.dumps(viewed_movies)
            
            outfile.write(string+"\n")
    return None

In [190]:
# Run this cell to test maskAsViewed
markAsViewed("It's a Beautiful Life")
markAsViewed("Black Panther")
markAsViewed("Thor")
# confirm that the viewed_movies list was built correctly
print(viewed_movies)
# confirm that the viewed_movies list was written to the file correctly
viewed_movies = loadViewedMovies()
print(viewed_movies)

["It's a Beautiful Life", 'Black Panther', 'Thor']
["It's a Beautiful Life", 'Black Panther', 'Thor']


<span style="color:blue">Expected output:</span><br/>
<span style="font: courier; font-size: 13px">
If the file viewed_movies.txt file was empty before you ran this cell, the expected output is shown below.<br />
["It's a Beautiful Life", 'Black Panther', 'Thor']<br />
["It's a Beautiful Life", 'Black Panther', 'Thor']</span>

### PROBLEM 8:
Define a function *getMovieAdvice* that takes the following steps:

1. Ask the user for a list of movies separated by commas
2. Call the sortedMoviesData function (defined earlier) on that list of movies (you need to split by commas)
3. Iterate over each movie. For each movie:
 - If the user has seen the movie before (use hasSeenMovie), print "You have already seen {title}"
 - If the user has NOT seen the movie before, print the title, year, rating, and plot in the format:

          {title} ({year}): {rating}
          --------------------
          {plot}

 - Ask the user if they want to watch the movie (yes or no)
      - If the user types "no" (or anything other than yes) do nothing.
      - If the user types "yes", then call markAsViewed on that movie
          title and print the runtime in the format:

          Watched {title} for {runtime}
4. return nothing

In [237]:
def getMovieAdvice():
    movielistOne=input("Please enter a list of movies separated by commas: ")
    movielist=movielistOne.split(",")
    result=sortedMoviesData(movielist)
    answeredyes=False
    for movie in result:
        if hasSeenMovie(movie)==True:
            print("You have already seen {}".format(movie))
        else:
            movieData=getMoviesData(movie)
            print('{} ({}): {}'.format(movie['Title'], movie['Year'],movie['imdbRating'])) 
            print('---------------------------')
            print(movie['Plot'])
            wanttowatch=input("Do you want to watch this movie? yes or no ")
            if wanttowatch=="yes" or wanttowatch=="YES" or wanttowatch=="Yes":
                markAsViewed(movie)
                print("Watched {} for {}".format(movie['Title'], movie['Runtime']))
                answeredyes=True
            if wanttowatch=="no" or wanttowatch=="NO" or wanttowatch=="No":
                None
            else:
                if answeredyes==True:
                    None
                else:
                    print("Invalid response")
            
    return None


In [238]:
getMovieAdvice()

Please enter a list of movies separated by commas: Lion King, Juno
The Lion King (1994): 8.5
---------------------------
Lion prince Simba and his father are targeted by his bitter uncle, who wants to ascend the throne himself.
Do you want to watch this movie? yes or noNo
Jenny, Juno (2005): 6.3
---------------------------
Fearing their families' disapproval, Jeni and her boyfriend Juno keep her pregnancy hidden. How long can the teenagers keep the secret?
Do you want to watch this movie? yes or nono


<span style="color:blue">Expected input and output:</span><br/>
<span style="font: courier; font-size: 13px">
Given the input:<br />
    Enter some movies separated by commas: Terminator,The Big Chill<br />
The expected interaction would be:<br />
The Big Chill (1983): 7.2<br />
--------------------<br />
A group of seven former college friends gather for a week-end reunion at a South Carolina vacation home after the funeral of another of their college friends.<br />
====================<br />
Want to watch this movie? (yes or no):yes<br />
Watched The Big Chill for 105 min<br />
Terminator (1991): 6.1<br />
--------------------<br />
A cyborg comes from the future, to kill a girl named Sarah Lee.<br />
====================<br />
Want to watch this movie? (yes or no):no</span>